docs/architecture/adr-016-protocol-versions.md

```diff
@@ -46,59 +44,64 @@ We need to do it in a way that is scalable and maintainable - we can't just litt
-  - Note the semantic meaning of a transaction may change according to the AppVersion, but the way txs are merklized into the header is part of the BlockVersion
+- All Tendermint hashed data-structures (headers, votes, txs, responses, etc.).
+    - Note the semantic meaning of a transaction may change according to the AppVersion, but the way txs are merklized into the header is part of the BlockVersion
 - It should be the least frequent/likely to change.
-  - Tendermint should be stabilizing - it's just Atomic Broadcast.
-  - We can start considering for Tendermint v2.0 in a year
+    - Tendermint should be stabilizing - it's just Atomic Broadcast.
+    - We can start considering for Tendermint v2.0 in a year
 - It's easy to determine the version of a block from its serialized form
 
 ### P2PVersion
 
-- All p2p and reactor messaging (messages, detectable behaviour)
+- All p2p and reactor messaging (messages, detectable behavior)
 - Will change gradually as reactors evolve to improve performance and support new features - eg proposed new message types BatchTx in the mempool and HasBlockPart in the consensus
 - It's easy to determine the version of a peer from its first serialized message/s
 - New versions must be compatible with at least one old version to allow gradual upgrades
 
 ### AppVersion
 
-- The ABCI state machine (txs, begin/endblock behaviour, commit hashing)
-- Behaviour and message types will change abruptly in the course of the life of a chain
+- The ABCI state machine (txs, begin/endblock behavior, commit hashing)
+- Behavior and message types will change abruptly in the course of the life of a chain
 - Need to minimize complexity of the code for supporting different AppVersions at different heights
-- Ideally, each version of the software supports only a _single_ AppVersion at one time
-  - this means we checkout different versions of the software at different heights instead of littering the code
-    with conditionals
-  - minimize the number of data migrations required across AppVersion (ie. most AppVersion should be able to read the same state from disk as previous AppVersion).
+- Ideally, each version of the software supports only a *single* AppVersion at one time
+    - this means we checkout different versions of the software at different heights instead of littering the code with conditionals
+    - minimize the number of data migrations required across AppVersion (ie. most AppVersion should be able to read the same state from disk as previous AppVersion).
+
+### RPCVersion
+
+- All rpc endpoints and their respective requests and responses.
+- Backwards compatibility should be upheld between minor software versions. i.e. we only add new endpoints for a minor software release 1.2 → 1.3 but can remove deprecated endpoints in a major release (1.5 → 2.0)
+- In the case of an introduction of a new interface with external users (i.e gRPC), this would also fall under the RPCVersion.
```

Comment:
The typical way to version RPC (at least with REST) is to use the major version as the base URL path, and never make backwards-incompatible changes to it. E.g. the URL would look something like:

```
http://node.domain.com:26657/v1/status
```

Any backwards-compatible changes to e.g. the `/status` endpoint we can just go ahead and make, while incompatible changes will need to go into `/v2/status`.

It may still make sense to include a minor version in e.g. headers so that clients can determine whether the server supports some feature which was introduced in a minor version.

Reply:
This makes sense. I will add a note in

Reply:
So still thinking in the mindset of hard forks, having this `/v1` and `/v2` won't work because a node that has been upgraded to v2 won't have data or won't be able to understand data from v1. If the node runner wanted to serve the data from the previous chain then they would just run a separate node to serve it. Would we still want to prepend the version number in the endpoint then?

Reply:
Yeah, that's a good question, and this really ties into the overall versioning strategy. I would hope that long-term we would like to never do hard-fork upgrades and instead support all historical block formats for the entire lifetime of a chain. But if we're going to be doing hard-forks all the time then this is probably not necessary, as long as we can always schedule RPC-breaking changes together with block-breaking changes.

Reply:
Yeah I'm also hoping something like that is possible

Reply:
I really like what Github does for versioning https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api (note: they don't use v1, vN as the base in URL path). why change the url when HTTP headers exist?

Reply:
Yup this is also an option. I think what I'm concerned about is that rpc clients need to find a way of knowing at which heights correlate with which versions. For example, if a block explorer doesn't know that an upgrade happens and it will look to fetch the block at the latest height (using v1) this won't be the true latest height which the block explorer would need to use v2 to find

Reply:
> I think what I'm concerned about is that rpc clients need to find a way of knowing at which heights correlate with which versions.

version can be inside the header, which is returned to the client

Reply:
only if the client didn't explicitly state the version right? then it would default to the latest version, and if the height the user had requested was for an earlier version then it would serve back that block and a lower version number so it could process the block accordingly

Reply:
right

Reply:
Note, I don't think there is any technical benefit to headers vs path -- both are very common.